File: leetcode/백트래킹 22. Generate Parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ret = []
        self.process(n, 0, 0, '', ret)
        return ret

    def process(self, n, openedCount, closedCount, s, ret):
        if openedCount == n and closedCount == n:
            ret.append(s)
            return

        if openedCount < n:
            self.process(n, openedCount + 1, closedCount, s + '(', ret)

        if openedCount > closedCount:
            self.process(n, openedCount, closedCount + 1, s + ')', ret)

File: leetcode/백트래킹 22. Generate Parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> ret = new ArrayList<>();
        recur(n, 0, 0, "", ret);
        return ret;
    }

    private void recur(int n, int openedCount, int closedCount, String str, List<String> ret) {
        if (openedCount == n && closedCount == n) {
            ret.add(str);
            return;
        }
        if (openedCount < n) {
            recur(n, openedCount + 1, closedCount, str + "(", ret);
        }
        if (openedCount > closedCount) {
            recur(n, openedCount, closedCount + 1, str + ")", ret);
        }
    }
}